pandora-common/src/assets/appearanceActionProcessingContext.ts

```diff
@@ -125,10 +125,24 @@ export class AppearanceActionProcessingContext {
 	}
 
 	public invalid(invalidReason?: InvalidActionReason): AppearanceActionProcessingResultInvalid {
-		return new AppearanceActionProcessingResultInvalid(this, invalidReason);
+		this.addProblem({
+			result: 'invalidAction',
+			reason: invalidReason,
+		});
+		return new AppearanceActionProcessingResultInvalid(this);
 	}
 
-	public finalize(): AppearanceActionProcessingResultValid {
+	public finalize(): AppearanceActionProcessingResultValid | AppearanceActionProcessingResultInvalid {
+		const validationResult = this.manipulator.currentState.validate();
+		if (!validationResult.success) {
+			this.addProblem({
+				result: 'validationError',
+				validationError: validationResult.error,
+			});
+		}
+		if (this._actionProblems.length > 0)
+			return new AppearanceActionProcessingResultInvalid(this);
+
 		return new AppearanceActionProcessingResultValid(this, this.manipulator.currentState);
```

Comment:
This change changes the behaviour of validation quite a lot.
The new behaviour means, that if there is any problem at all, the resulting state is not made available (even if the state itself is technically valid, just the transition isn't, e.g. due to blocked hands or permissions).
This removes the possibility of doing previews for actions that you can't do right now but are otherwise valid.
I'm not fully sure which way I prefer more, but I want to be really sure you intended to do this change.

Reply:
hmm, I made this change since I was debugging where the device store got invalid and it was super confusing to get a`ResultValid` class when the result was shown as not valid in the end

Reply:
I feel fine with not being able to show such previews, player can not get that state anyway so why show the preview